### DISCUSSION TITLE
PKI card works with opensc-pkcs11, but not with CTK (macos)

### DISCUSSION THREAD

**Discussion by the-docta** [OpenSC/OpenSC#3239] [Q&A]
I have a PKI card that contains a certificate used to access a certain web site (basically I authenticate myself using the card, or certificate from the card)

i'm on MacOS 15 now, tried opensc 0.25.1 (brew cask) and 0.26.0-rc1 (dmg from github), with same results

commands that work correctly:
```
pkcs15-tool --verify-pin 
pkcs11-tool --login --test
pcsctest
```

when I access the web site (chrome, edge, safari, firefox), the browser offers the certificate from the card, but never asks for the PIN. Access to web site is blocked then. firefox runs into timeout, chrome and edge show `ERR_SSL_CLIENT_AUTH_SIGNATURE_FAILED` safari just "can't open the page"

Neither the card nor the card reader are shown in the macos "keychain access" tool

what am I missing/ need to configure to force the browsers (or the OS) to ask me for the PIN

**Comment by mouse07410** [comment]
With Firefox, you may be in luck:
- go to Settings -> Privacy and Security;
- scroll down to the button "Security devices" and click on it;
- in the popup click "Load";
- type whatever name you choose, and in the path type /Library/OpenSC/lib/opensc-pkcs11.so
- click ok

From this point on, Firefox should work with your smart card. 

As for the other MacOS services and browsers - I've no idea. In theory, OpenSC includes OpenCTKToken.app or something like that (can't recall it's name, and don't want to bother looking) - it should make your token work with the rest of MacOS stuff. But it never worked for me, so I gave up on it years ago. 

If you're token/card is PIV- compliant, try `security list-smartcards`. If it shows that `pivtoken` started, you have a chance fur Safari, Chrome, etc.

**Reply by the-docta** [reply]
hi @mouse07410 thanks, that made firefox work. 

```
> security list-smartcards
org.opensc-project.mac.opensctoken.OpenSCTokenApp.OpenSCToken:101053060107
```

what does it mean?

**Reply by mouse07410** [reply]
In my understanding, it means that OpenSCToken got installed as a party of OpenSC and is running, but, like in my case before, does not work.
